File: 79. Word Search.py
class Solution(object):
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        visited = [[False for j in xrange(len(board[0]))] for i in xrange(len(board))]
        self.res = False
        for i in xrange(len(board)):
            for j in xrange(len(board[0])):
                if board[i][j] == word[0]:
                    self.dfs(i, j, 0, board, word, visited)
        return self.res
    
    def dfs(self, i, j, ind, board, word, visited):
        if ind == len(word):
            self.res = True
            return
        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or visited[i][j] or board[i][j] != word[ind]:
            return
        if not self.res:
            visited[i][j] = True
            self.dfs(i-1, j, ind+1, board, word, visited)
            self.dfs(i+1, j, ind+1, board, word, visited)
            self.dfs(i, j-1, ind+1, board, word, visited)
            self.dfs(i, j+1, ind+1, board, word, visited)
            visited[i][j] = False

File: 79. Word Search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @return {boolean}
 */
var exist = function(board, word) {
    let visited = [];
    for (let i = 0; i < board.length; i++)
    {
        visited.push([]);
        for (let j = 0; j < board[0].length; j++)
            visited[visited.length-1].push(false);
    }
    let res = false;
    var dfs = function(i, j, ind, visited)
    {
        if (ind == word.length)
        {
            res = true;
            return;
        }
        if (i < 0 || j < 0 || i >= board.length || j >= board[0].length || visited[i][j] || board[i][j] != word[ind])
            return;
        if (!res)
        {
            visited[i][j] = true;
            dfs(i-1, j, ind+1, visited);
            dfs(i+1, j, ind+1, visited);
            dfs(i, j-1, ind+1, visited);
            dfs(i, j+1, ind+1, visited);
            visited[i][j] = false;
        }
    };
    for (let i = 0; i < board.length; i++)
    {
        for (let j = 0; j < board[0].length; j++)
        {
            if (board[i][j] == word[0])
                dfs(i, j, 0, visited);
        }
    }
    return res;
};
